File: Data-Structures-and-Algorithms-using-Swift/String/Trie.swift

class Trie {
	class Node {
		var isLastChar : Bool = false
		var children: [Character:Node] = [:]
	}

	var root = Node()
	
	func insert(_ s : String) {
		if s == "" {
			return
		}
	
		let str : [Character] = Array(s) /*s.lowercased() is other option.*/
		self.root = self.insertUtil(node : self.root, str : str, index : 0)
	}
	
	func insertUtil(node : Node?, str : [Character], index : Int) -> Node {
		var curr : Node
		if node == nil {
			curr = Node()
		} else {
			curr = node!
		}
		
		if str.count == index {
			curr.isLastChar = true
		} else {
			curr.children[str[index]] = 
			self.insertUtil(node : curr.children[str[index]], str : str, index : index+1)
		}
		return curr
	}
	
	func remove(_ s : String) {
		if s == "" {
			return
		}
	
		let str : [Character] = Array(s.lowercased())
		self.removeUtil(curr : self.root, str : str,  index : 0)
	}
	
	func removeUtil(curr : Node?, str : [Character], index : Int) {
		guard let curr = curr else {
			return
		}
	
		if str.count == index {
			if curr.isLastChar {
				curr.isLastChar = false
			}
			return
		}
		self.removeUtil(curr : curr.children[str[index]], str : str, index : index+1)
	}
	
	func find(_ s : String) -> Bool {
		if s == "" {
			return false
		}
		let str : [Character] = Array(s) /* s.lowercased() */
		return self.findUtil(curr : self.root, str : str, index : 0)
	}
	
	func findUtil(curr : Node?, str : [Character], index : Int) -> Bool {
		guard let curr = curr else {
			return false
		}
		if str.count == index {
			return curr.isLastChar
		}
		return self.findUtil(curr : curr.children[str[index]], str : str, index : index+1)
	}
}

// Testing code
let tt = Trie()
tt.insert("banana")
tt.insert("apple")
tt.insert("mango")
print("Apple Found :", tt.find("apple"))
print("Banana Found :", tt.find("banana"))
print("Grapes Found :", tt.find("grapes"))

File: Data-Structures-Algorithms-In-TS/String/Trie.ts
let CharCount: number = 26;

class TrieNode {
    child: TrieNode[];
    isLastChar: boolean;
    ch: string;

    constructor(c: string) {
        this.child = new Array(CharCount).fill(null);
        this.isLastChar = false;
        this.ch = c;
    }
}

class Trie {
    root: TrieNode;

    constructor() {
        this.root = new TrieNode(' ');
    }

    insert(str: string): void {
        if (str == null) {
            return;
        }
        const temp = str.toString().toLowerCase();
        this.insertUtil(this.root, temp, 0);
    }

    private insertUtil(curr: TrieNode, str: string, index: number): TrieNode {
        if (curr == null) {
            curr = new TrieNode(str[index - 1]);
        }

        if (str.length === index) {
            curr.isLastChar = true;
        } else {
            curr.child[str.charCodeAt(index) - 'a'.charCodeAt(0)] =
                this.insertUtil(
                    curr.child[str.charCodeAt(index) - 'a'.charCodeAt(0)],
                    str, index + 1);
        }
        return curr;
    }

    remove(str: string): void {
        if (str == null) {
            return;
        }
        str = str.toLowerCase();
        this.removeUtil(this.root, str, 0);
    }

    private removeUtil(curr: TrieNode, str: string, index: number): void {
        if (curr == null) {
            return;
        }

        if (str.length === index) {
            if (curr.isLastChar) {
                curr.isLastChar = false;
            }
            return;
        }
        this.removeUtil(
            curr.child[str.charCodeAt(index) - 'a'.charCodeAt(0)],
            str,
            index + 1
        );
    }

    find(str: string): boolean {
        if (str == null) {
            return false;
        }
        str = str.toLowerCase();
        return this.findUtil(this.root, str, 0);
    }

    private findUtil(curr: TrieNode, str: string, index: number): boolean {
        if (curr == null) {
            return false;
        }
        if (str.length === index) {
            return curr.isLastChar;
        }
        return this.findUtil(
            curr.child[str.charCodeAt(index) - 'a'.charCodeAt(0)],
            str, index + 1);
    }
}

// Testing code.
const tt = new Trie();
tt.insert("banana");
tt.insert("apple");
tt.insert("mango");
console.log("Apple Found:", tt.find("apple"));
console.log("Grapes Found:", tt.find("grapes"));
console.log("Banana Found:", tt.find("banana"));

/*
Apple Found: true
Grapes Found: false
Banana Found: true
*/
